---
title: "02-scRNAseq-QC"
output:
  html_document:
    df_print: paged
date: "2024-02-19"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

## Intro

R version 4.3.1

Most of this code was either taken or adapted from code originally written by Deepti Murthy in the TCW Lab. I received assistancew from her and Alexandre Pelletier in writing this code. In this analysis, I remove cells with high mitochondrial percentage and remove outliers, scale, normalize, and cluster.

To begin, we load the necessary libraries into the R session:

```{r, echo = TRUE}
library(dplyr)
library(patchwork)
library(ggplot2)
library(gridExtra)
library(Matrix)
library(scales)
library(cowplot)
library(Seurat) #Seurat v.3.0
library(RColorBrewer)
library(BiocManager) #v 3.17
library(tibble)
library(patchwork)
library(flexmix)
library(miQC)
library(singleCellTK)
library(fgsea)
library(singleCellTK)
library(DropletUtils)
library(SeuratData)
library(tidyr)
```

## Step 1: Load in the Organoid Seurat Objects and Annotate with metadata

Second, we load in our Organoid Seurat objects as well as set directories identities. These objects underwent some preliminary QC in the previous step. Specifically, their ENSEMBL ID row names were converted to (mostly) gene symbols (I say mostly because there were some non-matches, which were left as the original ENSEMBL ID names). This is a large object, hence the need for this script. For context, this data is taken from 24 organoid samples (brain organoids), and the goal here is to accurately annotate each sample Seurat object with its corresponsing sample ID as well as its APOE genotype. The genotypes were discerned from the data sheet that came along with the sample data (meta data).

```{r, echo=TRUE}
setwd("/projectnb/tcwlab/LabMember/akg/APOE_Jorganoid_project/")

dir <- "/projectnb/tcwlab/LabMember/akg/APOE_Jorganoid_project/outputs"
out<-'outputs/simpleafSeurat'

S1 <- readRDS(file = "./outputs/simpleafSeurat/sample1.rds")
S2 <- readRDS(file = "./outputs/simpleafSeurat/sample2.rds")
S3 <- readRDS(file = "./outputs/simpleafSeurat/sample3.rds")
S4 <- readRDS(file = "./outputs/simpleafSeurat/sample4.rds")
###* No Sample 5 (deemed poor quality by researchers)
S6 <- readRDS(file = "./outputs/simpleafSeurat/sample6.rds")
###* No Sample 7 (deemed poor quality by researchers)
S8 <- readRDS(file = "./outputs/simpleafSeurat/sample8.rds")
S9 <- readRDS(file = "./outputs/simpleafSeurat/sample9.rds")
S10 <- readRDS(file = "./outputs/simpleafSeurat/sample10.rds")
S11 <- readRDS(file = "./outputs/simpleafSeurat/sample11.rds")
S12 <- readRDS(file = "./outputs/simpleafSeurat/sample12.rds")
S13 <- readRDS(file = "./outputs/simpleafSeurat/sample13.rds")
S14 <- readRDS(file = "./outputs/simpleafSeurat/sample14.rds")
S15 <- readRDS(file = "./outputs/simpleafSeurat/sample15.rds")
S16 <- readRDS(file = "./outputs/simpleafSeurat/sample16.rds")
S17 <- readRDS(file = "./outputs/simpleafSeurat/sample17.rds")
S18 <- readRDS(file = "./outputs/simpleafSeurat/sample18.rds")
S19 <- readRDS(file = "./outputs/simpleafSeurat/sample19.rds")
S20 <- readRDS(file = "./outputs/simpleafSeurat/sample20.rds")
S21 <- readRDS(file = "./outputs/simpleafSeurat/sample21.rds")
S22 <- readRDS(file = "./outputs/simpleafSeurat/sample22.rds")
S23 <- readRDS(file = "./outputs/simpleafSeurat/sample23.rds")
S24 <- readRDS(file = "./outputs/simpleafSeurat/sample24.rds")
S25 <- readRDS(file = "./outputs/simpleafSeurat/sample25.rds")
S26 <- readRDS(file = "./outputs/simpleafSeurat/sample26.rds")

###* Proceed by merging all the samples into a single object, called organoid0
APOE22_samples_group <- c("S3", "S9", "S12", "S15", "S21", "S25")
APOE33_samples_group <- c("S1", "S8", "S14", "S16", "S19", "S23")
APOE33Ch_samples_group <- c("S2", "S6", "S13", "S18", "S20", "S24")
APOE44_samples_group <- c("S4", "S10", "S11", "S17", "S22", "S26")

sample_names_group22 <- c("sample3", "sample9", "sample12", "sample15", "sample21", "sample25")
sample_names_group33 <- c("sample1", "sample8", "sample14", "sample16", "sample19", "sample23")
sample_names_group33Ch <- c("sample2", "sample6", "sample13", "sample18", "sample20", "sample24")
sample_names_group44 <- c("sample4", "sample10", "sample11", "sample17", "sample22", "sample26")
```

Now that we have loaded in our 24 Seurat objects, we can proceed by labeling each of them by their respective APOE genotypes and sample IDs:

```{r, echo=TRUE}
# Loop through each sample and add a metadata column for sample labels
# for APOE22
for (i in seq_along(APOE22_samples_group)) {
  object_name <- APOE22_samples_group[i]
  sample_name <- sample_names_group22[i]
  
  eval(parse(text = paste0(object_name, "[['sample']] <- '", sample_name, "'")))
  eval(parse(text = paste0(object_name, "[['genotype']] <- 'APOE22'")))
}
# APOE33
for (i in seq_along(APOE33_samples_group)) {
  object_name <- APOE33_samples_group[i]
  sample_name <- sample_names_group33[i]
  
  eval(parse(text = paste0(object_name, "[['sample']] <- '", sample_name, "'")))
  eval(parse(text = paste0(object_name, "[['genotype']] <- 'APOE33'")))
}
# APOE33Christchurch line
for (i in seq_along(APOE33Ch_samples_group)) {
  object_name <- APOE33Ch_samples_group[i]
  sample_name <- sample_names_group33Ch[i]
  
  eval(parse(text = paste0(object_name, "[['sample']] <- '", sample_name, "'")))
  eval(parse(text = paste0(object_name, "[['genotype']] <- 'APOE33Ch'")))
}
# APOE44
for (i in seq_along(APOE44_samples_group)) {
  object_name <- APOE44_samples_group[i]
  sample_name <- sample_names_group44[i]
  
  eval(parse(text = paste0(object_name, "[['sample']] <- '", sample_name, "'")))
  eval(parse(text = paste0(object_name, "[['genotype']] <- 'APOE44'")))
}

```


## Step 2: Combine the Individual Seurat Objects into a Single Seurat Object

Now we will combine all of these individually-processed Seurat objects into one Seurat object, which we will use to do the rest of our QC. At this point, it would also be prudent to use rm() to remove all of the previous Seurat objects from the R working memory.

```{r, echo=TRUE}
organoid <- merge(S1, y = c(S2, S3, S4, S6, S8, S9, S10,
                             S11, S12, S13, S14, S15, S16, S17, S18, S19,
                             S20, S21, S22, S23, S24, S25, S26), 
                   add.cell.ids = c("S1", "S2", "S3", "S4", "S6", "S8", "S9", "S10",
                                    "S11", "S12", "S13", "S14", "S15", "S16", "S17", "S18", "S19",
                                    "S20", "S21", "S22", "S23", "S24", "S25", "S26"),
                   project = "APOE_Jorganoid"
)

out <- "/projectnb/tcwlab/LabMember/akg/APOE_Jorganoid_project/outputs/02-scRNAseq-QC/"
filename = 'organoid.rds'
saveRDS(organoid, file.path(out, filename))

rm(S1, S2, S3, S4, S6, S8, S9, S10,
   S11, S12, S13, S14, S15, S16, S17, S18, S19, S20,
   S21, S22, S23, S24, S25, S26)
```

## Step 3: Identify mtDNA Genes and Perform Mitochondrial QC

We start this section by defining a list of all 37 Mitochondrial DNA genes' ENSEMBL IDs (the coreesponding gene symbols are in the comments). Note that this is being done because the mtDNA genes were not converted from ENSEMBL to gene symbol in the previous RMarkDown's code. So, to make up for this, we do this manual annotation. Note: this code may take a long time to execute.

```{r, echo=TRUE}
### all 37 Human Mitochondrial genes
mt_gene_names <- c(
  "ENSG00000198888", # ND1
  "ENSG00000198763", # ND2
  "ENSG00000198804", # COX1
  "ENSG00000198712", # COX2
  "ENSG00000228253", # ATP8
  "ENSG00000198899", # ATP6
  "ENSG00000198938", # COX3
  "ENSG00000198840", # ND3
  "ENSG00000212907", # ND4L
  "ENSG00000198886", # ND4
  "ENSG00000198786", # ND5
  "ENSG00000198695", # ND6
  "ENSG00000198727", # CYTB
  "ENSG00000210049", # MT-TF
  "ENSG00000210082", # MT-TV
  "ENSG00000209082", # MT-TL1
  "ENSG00000198888", # MT-TR
  "ENSG00000210100", # MT-TN
  "ENSG00000210107", # MT-TG
  "ENSG00000210112", # MT-TL2
  "ENSG00000210117", # MT-TS1
  "ENSG00000210127", # MT-TV
  "ENSG00000210135", # MT-TE
  "ENSG00000210144", # MT-TS2
  "ENSG00000210151", # MT-TH
  "ENSG00000210154", # MT-TD
  "ENSG00000210156", # MT-TK
  "ENSG00000210164", # MT-TM
  "ENSG00000210174", # MT-TI
  "ENSG00000210184", # MT-TT
  "ENSG00000210191", # MT-TW
  "ENSG00000210194", # MT-TC
  "ENSG00000210211", # MT-TY
  "ENSG00000210228", # MT-TA
  "ENSG00000210243", # MT-TQ
  "ENSG00000198899", # MT-RNR1 (12S rRNA)
  "ENSG00000198763",  # MT-RNR2 (16S rRNA)
  "ND1", "ND2", "COX1", "COX2", "ATP8", "ATP6", "COX3", 
  "ND3", "ND4L", "ND4", "ND5", "ND6", "CYTB",
  "TF", "TV", "TL1", "TR", "TN", "TG", "TL2", "TS1", "TV",
  "TE", "TS2", "TH", "TD", "TK", "TM", "TI", "TT", "TW", 
  "TC", "TY", "TA", "TQ", "RNR1", "RNR2"
)

# Initialize a data frame to store cell counts before and after QC
cell_counts <- data.frame(sample = character(), 
                          cells_before_QC = integer(), 
                          cells_after_QC = integer())

# Get unique sample identifiers
samples <- organoid@meta.data$sample

# Define mitochondrial genes (adjust according to your organism)
mt_gene_names <- grep("^MT-", rownames(organoid), value = TRUE)
genotypes <- organoid@meta.data$genotype

# Loop over each sample
for (sample_id in samples) {
  # Subset Seurat object for the current sample
  QC <- subset(organoid, subset = sample == sample_id)
  genotype <- genotypes[i]
  # Record the number of cells before QC
  cells_before <- ncol(QC)
  
  # Filter genes expressed in less than 3 cells
  raw_counts_mat <- QC@assays$RNA@counts
  genes_to_keep <- Matrix::colSums(raw_counts_mat != 0) >= 3
  QC <- subset(QC, features = rownames(QC)[genes_to_keep])
  
  # Calculate upper limit, set at 99.5 percentile
  upper_limit <- quantile(QC@meta.data$nCount_RNA, probs = 0.995)
  
  QC <- CalculateBarcodeInflections(
    QC,
    barcode.column = "nCount_RNA",
    group.column = "orig.ident",
    threshold.low = 1000,
    threshold.high = NULL
  )
  QC <- SubsetByBarcodeInflections(object = QC)
  
  # Record the number of cells after QC
  cells_after <- ncol(QC)
  
  # Append the cell count information to the data frame
  cell_counts <- rbind(cell_counts, data.frame(sample = sample_id,
                                               cells_before_QC = cells_before,
                                               genotype = genotype,
                                               cells_after_QC = cells_after))
}

aggregated_cell_counts <- aggregate(. ~ genotype, data = cell_counts, FUN = sum)

# View the cell counts data frame
#print(cell_counts)
print(aggregated_cell_counts)
```


Now that we have removed cells with either too much mtDNA (likely cells that are damaged, lysed, or undergoing cellular stress), we can proceed with the Ambient RNA correction.

## Step 4: Ambient RNA Correction

We opted for the SoupX package in this case. This is because the SoupX package refers to the raw counts matrices when performing the analysis in removing empty droplets. It's a powerful tool that we are able to use because we have the raw counts directories. The first step will be to identify the locations of each of these raw counts directories, as well as define a function (used in the previous RMarkDown notebook) which will convert ENSEMBL IDs to gene symbols.

```{r, echo = TRUE}

sample_ids <- unique(organoid@meta.data$sample)
# sample_ids
seurat_objects <- SplitObject(organoid, split.by = "sample")

# Load raw counts
custom_format <- list("counts" = c("U","S","A"))
# Directories for raw counts of each sample
raw_counts_directories <- c("1_S16/af_quant/", "2_S8/af_quant/", "3_S9/af_quant/", "4_S14/af_quant/", "6_S23/af_quant/", "8_S17/af_quant/", "9_S21/af_quant/", "10_S7/af_quant/", "11_S1/af_quant/", "12_S20/af_quant/", "13_S11/af_quant/", "14_S13/af_quant/", "15_S5/af_quant/", "16_S24/af_quant/", "17_S19/af_quant/", "18_S22/af_quant/", "19_S3/af_quant/", "20_S15/af_quant/", "21_S6/af_quant/", "22_S2/af_quant/", "23_S12/af_quant/", "24_S10/af_quant/", "25_S18/af_quant/", "26_S4/af_quant/") 
# paths to quant files
### Function to convert ENSEMBL_ID names to gene symbols
getGeneSymbols <- function(ensembl_ids) {
  require(org.Hs.eg.db)
  gene_symbols <- mapIds(org.Hs.eg.db,
                         keys = ensembl_ids,
                         column = "SYMBOL",
                         keytype = "ENSEMBL",
                         multiVals = "first")
  return(gene_symbols)
}


```

We'll also want to initialize a dataframe which will contain all of our SoupX RNA counts so we can track the before and after QC metrics:

```{r, echo = TRUE}
# initialize a dataframe which will contain the before/after SoupX correction data
SoupXSummary <- data.frame(
  sampleID = character(24),
  before_SoupX = numeric(24),
  after_SoupX = numeric(24),
  percent_removed = numeric(24),
  stringsAsFactors = FALSE
)

```

Then, we will define our SoupX correction function:

```{r, echo = TRUE}
source <- '/projectnb/tcwlab/LabMember/akg/APOE_Jorganoid_project/outputs/'
# Function to apply SoupX correction
apply_soupX_correction <- function(seurat_obj, raw_counts_directory) {
  
  # # set a "raw counts" count for each of the individual Seurat samples' objects
  # seurat_obj[["original_counts"]] <- CreateAssayObject(counts = seurat_obj@assays$RNA@counts) #raw counts
  # 
  sce <- fishpond::loadFry(raw_counts_directory,
                           outputFormat = custom_format)
  # Assuming 'sce' is your Single Cell Experiment object
  # Replace Ensembl IDs in row names with gene symbols
  ensembl_ids_sce <- rownames(sce)
  gene_symbols_sce <- getGeneSymbols(ensembl_ids_sce)
  
  # Replace Ensembl IDs with gene symbols in row names
  # If no gene symbol is found, retain the original Ensembl ID
  updated_rownames <- ifelse(is.na(gene_symbols_sce[ensembl_ids_sce]), ensembl_ids_sce, gene_symbols_sce[ensembl_ids_sce])
  rownames(sce) <- updated_rownames
  
  #create the seurat object by filtering for selected cells
  S1Counts <- counts(sce)
  
  raw <- S1Counts
  #raw counts
  seurat_obj[["original_counts"]] <- CreateAssayObject(counts = seurat_obj@assays$RNA@counts) 
  
  #SoupX Correction
  #modify genes in raw
  genes <- intersect(rownames(seurat_obj), rownames(raw))
  length(genes)
  raw <- raw[genes,]
  
  #run SoupX algo (this package is what I ended up using instead of DecontX)
  sc <- SoupChannel(raw, seurat_obj@assays$RNA@counts[genes,])
  sc <- setClusters(sc, seurat_obj$seurat_clusters)
  sc <- autoEstCont(sc, doPlot = FALSE)
  soup_out <- adjustCounts(sc, roundToInt = TRUE) #adjusted counts matrix
  
  seurat_obj[["SoupX_counts"]] <- CreateAssayObject(counts = soup_out)
  seurat_obj@assays$RNA@counts <- soup_out #USED
  
  #calculate percent of mRNA removed
  percent_removed_soup <- (1 - (sum(seurat_obj@assays$RNA@counts) / sum(seurat_obj@assays$original_counts@counts)))*100
  seurat_obj[["percent_mRNA_removed"]] <- percent_removed_soup
  
  return(seurat_obj)
}

```

The final steps will be to 1. loop through each of the Seurat objects, applying the SoupX correction to each, 2. quantifying those results, and 3. merging the objects back into one combined Seurat object, which will be saved and used for the next script's analysis. Steps 1 and 2:

```{r, echo = TRUE}
# Process each Seurat object
for (i in seq_along(seurat_objects)) {
  sampleID <- names(seurat_objects)[i]
  print(sampleID)
  raw_counts_dir <- paste(source, raw_counts_directories[i], sep = '') 
  print(raw_counts_dir)
  original_cell_count <- ncol(seurat_objects[[sampleID]]@assays$RNA@counts)
  
  seurat_objects[[sampleID]] <- apply_soupX_correction(seurat_objects[[sampleID]], raw_counts_dir)
  
  after_cell_count <- ncol(seurat_objects[[sampleID]]@assays$RNA@counts)
  percent_removed <- (1 - (after_cell_count / original_cell_count)) * 100
  
  SoupXSummary[i, ] <- c(sampleID, original_cell_count, after_cell_count, percent_removed)
}

### this is to get the total mRNAs, removed mRNAs, percent removed mRNAs from each sample
for (i in seq_along(seurat_objects)) {
  sampleID <- names(seurat_objects)[i]
  print(sampleID)
  print(unique(seurat_objects[[i]]@meta.data$percent_mRNA_removed))
  
  print(sum(seurat_objects[[i]]@assays$original_counts@counts))
  print((sum(seurat_objects[[i]]@assays$RNA@counts)))
}


```

Finally, we save combined object:

```{r}
organoid <- merge(seurat_objects[[1]], y = c(seurat_objects[2:length(seurat_objects)]),
                   project = "APOE_Jorganoid"
)

saveRDS(organoid,file.path(out,'organoid.rds'))
```